infrastructure/kubernetes/challenges/README.md

```diff
@@ -0,0 +1,4 @@
+# Challenge directory
+This directory will host all challenges.
```

Comment:
Hmm, I don't really like that the challenges are inside the infrastructure repo. I think this should be separate.
I.e. I think it would be nice to just add the kubernetes (or kubernetes/bin) folder to your path and run the scripts from whereever. In that case, it would be nice to have a common prefix to easily find them with tab completion. E.g. kctf-create-cluster.sh

Reply:
What about a venv-style setup? You call a script that gives you a shell with the right tools in the path, and the right directory with the challenges.

Say:
  ./scripts/setup/activate.sh some-directory

Then some-directory/config/... will hold the config and some-directory/challenges/... will hold the tasks and some-directory/.src will symlink to the code.

I like that a bit more than a generic PATH script because then the scripts can always find not only the challenges, but also other things (config, template, etc). Since not all things are challenge-driven (eg, making a cluster), we can't always use the challenge path without making assumptions about the directory structure anyway, so this way we make it more obvious.

Reply:
But you know the config directory path already based on the script location and the challenge path should come from the argument (+ cwd).
Usually the challenges will be in a separate git project than the infrastructure scripts and not all challenges might be using this cluster.

Also I think it would be nice to not make this too magical. I.e. the challenge directory should work standalone with kubernetes only even if you don't have the infrastructure scripts installed. That way it will be easy for people to administer a challenge if they have some k8s knowledge without knowing anything about the infrastructure tools themselves.

Reply:
Yea, it'll always be possible to run things with k8s directly :-) that would never change. The question is more about the experience we expect to provide to organizers, and what use cases become impossible with certain decisions.

The problem is that there's an explicit relationship between cluster to challenge, and the configuration per cluster can't live in challenge directories. If the configuration of the cluster lives in the infrastructure repo, then why shouldn't the challenges live there too? And viceversa. It would be more consistent if the cluster configuration was treated the same way as the challenge details.

A more concrete example could be, let's say you organize two events in a year, I imagine people would end up with 6 clusters.
  1. Testrun / Development for quals
  2. Qualifications
  3. Post-competition hosting of the quals
  4. Testrun / Development for finals
  5. Finals
  6. Post-competition hosting of the finals

The challenges 1-3 are pretty much gonna be the same (same for 4-6), but it would be beneficial to keep the configuration of each cluster stored somewhere. If the configuration lives in ../../config then it's impossible to manage them independently (specially since cluster #3 will continue to run and might need to be updated while 4-6 are done, and same for cluster 1 running during #2 and #4 running during #5).

As a user, what I would expect be the easiest would be:
  1. Create cluster #1 and then all the challenges (the challenges directory itself can be a symlink, or the tasks could be the symlink).
  2. Create cluster #2 and then symlink some tasks from #1/challenges/... to #2/challenges/...
  3. Create cluster #3 and then symlink the full challenges directory from #2.
  4. For #4-6 the same would apply.

I agree that having the tasks themselves in the same directory as git would be bad, but we can't unlink cluster-challenges completely, there has to be a link in both directions (so actions on the task can consider the cluster, and so you can make cluster-wide actions).

So, I think that having the config in ../../config and the challenges in random directories across the FS would lead to:
  1. Making it impossible for a script acting on a specific challenge to know which cluster to modify.
  2. Making it impossible for a script acting on a cluster to make actions that affect all tasks.

So, I think taking a similar approach to git/venv (where some directory is taken over as a workspace) would easily solve those concerns without any downsides?

Reply:
I don't understand why you need a connection from challenge to cluster.
What I would want is a way to set a cluster as active and then whenever I touch a challenge it should use the current cluster.

Reply:
Maybe the issue you are asking about is the connection from cluster to challenge (not the other way around). That's the second point from my previous response (that design makes it impossible to build scripts that affects all tasks of a cluster). That's obvious (I mean, if the tasks can be all around the place, it's impossible for a script to know where they are).

Reply:
For context, the current (internal) infra has several actions that apply to all tasks. If we don't make that part of this, then everyone will probably have to build one themselves.